### ISSUE TITLE
ncep_classes.py updates

### ISSUE THREAD

**Issue by flampouris** [JCSDA/ioda-converters#204]
A couple of updates are needed for the ncep_classes.py
 
- Remove the prepbufr/ bufr separation; it does not offer something extra as long, the bufr tables are used;

- Generalize the use of the message name, e.g., NC001002 or ADPUPA or radiosonde, etc. The fast update instead of generalization: There is an if statement to select a different kind, currently this does not fail (I have to spend some more time with this);

- Make clear (as required input argument) when the intespec.yaml is provided and when the ncep_classes.py creates it;

- All the if statements should be based on bufr table, remove the NetCDF variable names from the if statements;

- Stop supporting python2, e.g. DateTime module has differences between python2 and python3;

- Update the writer to use the interface and C++;

There are will be several PRs and this issue will be updated accordingly.

**Comment by srherbener** [comment]
@flampouris, is this issue closed with the recent updates to ncep_classes.py? If it is agreeable with you, the parts of this issue that are not finished yet could be listed in new issues (one per task), and then close this one. Thanks!

**Comment by flampouris** [comment]
@srherbener These are still open issues that I did not have the time to address, except the python2/3 problem at feature/ncep_class_updates branch. 

I do not know what @emilyhcliu is planning to do with the ioda-converters, but I hope that we can coordinate to address them. 

Otherwise we can close the issue and I will reopen it when I have some time.

**Comment by srherbener** [comment]
@flampouris and @emilyhcliu, let's leave this open for now until you get your plans squared away. I want to make sure we remember to do this work (whenever it makes sense to do so).

**Comment by srherbener** [comment]
@flampouris, @emilyhcliu what is the status of this work? Can this issue be closed? Thanks!

**Comment by emilyhcliu** [comment]
@srherbener @flampouris I did not notice this issue was assigned to me until today. So I did not respond until today.  I apologize for my late response.  My use of ioda-converter is mainly to generate the test data sets from GSI NC diagnostic files.  The problem in this issue is related to the conversion of bufr to IODA format.  I have not run into a chance to work with ncep_classes.py before, so I do not have a sense of the improvements need to be done.   I know that the NCEP is organizing the observation processing team to work with the JEDI IODA team.  I suggest we close this issue for now.  
If there is a need to improve the ncep_classes.py during the collaboration between NCEP and IODA, I believe the issue will be raised again.

**Comment by srherbener** [comment]
Thanks @emilyhcliu! @flampouris is it okay with you to  close this. We can submit specific ZenHub issues in the future if anything comes up. Thanks!

**Comment by flampouris** [comment]
@srherbener

 @YanHao-NOAA is currently working on the updates of the ncep_classes. Hopefully, there will be a PR in the next days or so. 

I update the assignment.

**Comment by srherbener** [comment]
@flampouris, thanks for the response. I'll move this issue to the "In Progress" stage.

**Comment by srherbener** [comment]
@YanHao-NOAA, does your current PR (#296) address everything remaining in this issue? If so, please attach this issue to PR #296. Also, in the initial description in PR #296, add the text `Closes #204` and when the PR gets merged, this issue will automatically get closed.

Thanks!

**Comment by YanHao-NOAA** [comment]
@srherbener I did not notice this issue.
No, not all the problems in this issue were solved. 
- Remove the prepbufr/ bufr separation; it does not offer something extra as long, the bufr tables are used; (Not solved)

- Generalize the use of the message name, e.g., NC001002 or ADPUPA or radiosonde, etc. The fast update instead of generalization: There is an if statement to select a different kind, currently this does not fail (I have to spend some more time with this) ( Solved)

Make clear (as required input argument) when the intespec.yaml is provided and when the ncep_classes.py creates it; (Solved)

All the if statements should be based on bufr table, remove the NetCDF variable names from the if statements; (Not solved)

Stop supporting python2, e.g. DateTime module has differences between python2 and python3; (Solved)

Update the writer to use the interface and C++; (Not sure about this one)

Thanks!
Yan

**Comment by srherbener** [comment]
Thanks for the response @YanHao-NOAA. At this point, I think it would be good to close this issue, and submit new issues, one each, to cover the bullet points that are not solved yet. Doing this will help us track and schedule these updates more effectively.

Could you do this? I think it would work to create two new issues, one each for the items you marked "Not solved". Let's not create an issue for the "Update the writer to use the interface and C++" since this will be covered in our upcoming effort to migrate the converters to the new idoa interface. 

Thanks!

**Comment by YanHao-NOAA** [comment]
@srherbener , I think it is a good idea to close this issue. I will submit two new issues.

- Issue one : "remove the prepbufr/ bufr separation", I have a question about this issue: are we planning to drop everything related to prepbufr in this code?  

- Issue two: "if statements should be based on bufr table, remove NetCDF variables", I will look into this one, not sure which variables were referred here.

Thanks!

**Comment by srherbener** [comment]
@YanHao-NOAA, thanks for looking into this. I think the contents of your code should be driven by the people that will be using your converter. I'm not the one to decide if you should or should not include prepbufr.

It might be good to ask Stelios these questions since he was the one who originally submitted this issue.

Thanks!

**Comment by flampouris** [comment]
Oh! That's from last August...

- There is no real reason to invest time on the first issue, it's a simplification for the user. 
- I cannot recall the reason for the second issue, probably it has to do with some shortcut that I took during development.

If it was my call,  I would close this issue. Especially if there are plans for the new ioda engine and obs in BUFR. @srherbener apologies, I did not manage to attend your presentation about the ioda engines last week.
